File: backend/src/main/java/com/gargujjwal/military_asset_management/service/EquipmentService.java
package com.gargujjwal.military_asset_management.service;

import com.gargujjwal.military_asset_management.dto.EquipmentCategoryDetailDto;
import com.gargujjwal.military_asset_management.dto.EquipmentCategoryDto;
import com.gargujjwal.military_asset_management.dto.EquipmentDetailDto;
import com.gargujjwal.military_asset_management.dto.EquipmentDto;
import com.gargujjwal.military_asset_management.entity.Equipment;
import com.gargujjwal.military_asset_management.entity.EquipmentCategory;
import com.gargujjwal.military_asset_management.exception.ResourceNotFoundException;
import com.gargujjwal.military_asset_management.mapper.EquipmentMapper;
import com.gargujjwal.military_asset_management.repository.EquipmentCategoryRepository;
import com.gargujjwal.military_asset_management.repository.EquipmentRepository;
import java.util.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Slf4j(topic = "EQUIPMENT_SERVICE")
@RequiredArgsConstructor
public class EquipmentService {
  private final EquipmentRepository equipmentRepository;
  private final EquipmentCategoryRepository equipmentCategoryRepository;
  private final EquipmentMapper equipmentMapper;

  public EquipmentDetailDto getEquipmentDetailById(String id) {
    return equipmentMapper.toDetailDto(findEquipmentById(id));
  }

  public List<EquipmentCategoryDetailDto> getAllEquipmentCategories() {
    return equipmentMapper.toDetailDto(equipmentCategoryRepository.findAll());
  }

  @PreAuthorize("hasRole('ROLE_ADMIN')")
  @Transactional
  public void createEquipmentCategory(EquipmentCategoryDto newEquipmentCategory) {
    EquipmentCategory equipmentCategory = equipmentMapper.toEntity(newEquipmentCategory);
    equipmentCategory.setId(null);
    equipmentCategoryRepository.save(equipmentCategory);
  }

  @PreAuthorize("hasRole('ROLE_ADMIN')")
  @Transactional
  public void createEquipment(EquipmentDto newEquipment, String categoryId) {
    Equipment equipment = equipmentMapper.toEntity(newEquipment);
    equipment.setId(null);
    EquipmentCategory equipCategory = findCategoryById(categoryId);

    // for hibernate
    equipment.setEquipmentCategory(equipCategory);

    equipmentRepository.save(equipment);
  }

  @PreAuthorize("hasRole('ROLE_ADMIN')")
  @Transactional
  public void deleteEquipmentCategory(String equipmentCategoryId) {
    findCategoryById(equipmentCategoryId);
    equipmentCategoryRepository.deleteById(equipmentCategoryId);
  }

  @PreAuthorize("hasRole('ROLE_ADMIN')")
  @Transactional
  public void deleteEquipment(String equipmentId) {
    findEquipmentById(equipmentId);
    equipmentRepository.deleteById(equipmentId);
  }

  private EquipmentCategory findCategoryById(String id) {
    return equipmentCategoryRepository
        .findById(id)
        .orElseThrow(() -> new ResourceNotFoundException("Equipment category not found"));
  }

  private Equipment findEquipmentById(String id) {
    return equipmentRepository
        .findById(id)
        .orElseThrow(() -> new ResourceNotFoundException("Equipment not found"));
  }
}

File: frontend/app/services/equipment-service.ts
import type {
  EquipmentCategoryDetailDto,
  EquipmentCategoryDto,
  EquipmentDetailDto,
  EquipmentDto,
} from "~/types/backend-stubs";
import { fetchWithAuth } from "~/utils/api";

export function getAllEquipmentCategories() {
  return fetchWithAuth<EquipmentCategoryDetailDto[]>("/equipments/categories");
}

export function getEquipmentDetailById(id: string) {
  return fetchWithAuth<EquipmentDetailDto>(`/equipments/${id}`);
}

export function createEquipmentCategory(
  newEquipmentCategory: EquipmentCategoryDto
) {
  return fetchWithAuth<null>("/equipments/categories", {
    body: JSON.stringify(newEquipmentCategory),
    method: "POST",
  });
}

export function createEquipment(
  categoryId: string,
  newEquipment: EquipmentDto
) {
  return fetchWithAuth<null>(
    `/equipments/categories/${categoryId}/equipments`,
    {
      body: JSON.stringify(newEquipment),
      method: "POST",
    }
  );
}

export function deleteEquipmentCategory(id: string) {
  return fetchWithAuth<null>(`/equipments/categories/${id}`, {
    method: "DELETE",
  });
}

export function deleteEquipment(id: string) {
  return fetchWithAuth<null>(`/equipments/${id}`, {
    method: "DELETE",
  });
}
